### PR TITLE
refactor(frontend): improve readability of LogicalSource

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

`LogicalSource` is quite hard to understand, because it does too many things.

This PR removes a lot of unnecessary functions to make it more straightforward.
- replace `gen_row_id` and `for_table` with `kind`
- remove `wrap_with_optional_generated_columns`, and `dispatch_new_s3_plan`. Create plan nodes directly. 

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #14687** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14687?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #14560** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14560?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #14171** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14171?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #14172** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14172?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **PR #14465** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14465?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/risingwavelabs/risingwave/14687?utm_source=stack-comment).

File: src/frontend/src/optimizer/plan_node/logical_source.rs [hunk 1]
```diff
@@ -85,11 +85,13 @@ impl LogicalSource {
         let base = PlanBase::new_logical_with_core(&core);
 
         let output_exprs = Self::derive_output_exprs_from_generated_columns(&core.column_catalog)?;
+        let (core, output_row_id_index) = core.exclude_generated_columns();
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/logical_source.rs:88]: I move the logic in `rewrite_to_stream_batch_source` here, which can help simplify the logic in `to_stream` a lot.

File: src/frontend/src/optimizer/plan_node/generic/source.rs [hunk 2]
```diff
@@ -28,20 +29,37 @@ use crate::optimizer::optimizer_context::OptimizerContextRef;
 use crate::optimizer::property::FunctionalDependencySet;
 use crate::{TableCatalog, WithOptions};
 
+/// In which scnario the source node is created
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+#[expect(clippy::enum_variant_names)]
+pub enum SourceNodeKind {
```
Comment (tabVersion) [inline src/frontend/src/optimizer/plan_node/generic/source.rs:35]: can you add another field to distinguish cdc source?

File: src/frontend/src/optimizer/plan_node/generic/source.rs [hunk 3]
```diff
@@ -80,6 +98,30 @@ impl GenericPlanNode for Source {
...
 impl Source {
+    pub fn is_new_fs_connector(&self) -> bool {
+        self.catalog.as_ref().is_some_and(|catalog| {
+            ConnectorProperties::is_new_fs_connector_b_tree_map(&catalog.with_properties)
+        })
+    }
+
+    /// The columns in stream/batch source node indicate the actual columns it will produce,
+    /// instead of the columns defined in source catalog. The difference is generated columns.
+    pub fn exclude_generated_columns(mut self) -> (Self, Option<usize>) {
```
Comment (tabVersion) [inline src/frontend/src/optimizer/plan_node/generic/source.rs:109]: off topic, how source backfill handles generated columns

**Review by tabVersion — COMMENTED** [review]
generally LGTM, thanks for refactoring.

File: src/frontend/src/optimizer/plan_node/generic/source.rs (hunk 2 above)
Comment (xxchan) [reply src/frontend/src/optimizer/plan_node/generic/source.rs:35]: I think cdc source can just use CreateSourceWithStreamjob

File: src/frontend/src/optimizer/plan_node/generic/source.rs (hunk 3 above)
Comment (xxchan) [reply src/frontend/src/optimizer/plan_node/generic/source.rs:109]: Put Project after backfill node.

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**